File: 0047-permutations-ii/0047-permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    const memo = {}
    nums.sort((a, b) => a - b)
    return helper(nums)
    
    function helper(nums) {
        if (nums.length < 2) {
            return memo[nums] = [nums]
        }
        if (nums in memo) {
            return memo[nums]
        }
        const res = []
        for (let i=0; i < nums.length; i++) {
            if (i > 0 && nums[i] === nums[i-1]) continue
            const num = nums[i]
            const rest = helper(nums.slice(0, i).concat(nums.slice(i+1)))
            for (const r of rest) {
                res.push([num, ...r])
            }
        }
        return memo[nums] = res
    }
};

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        
        def permute(nums: List[int]) -> List[List[int]]:
            if len(nums) == 1:
                return [nums]
            res = []
            for i in range(0, len(nums)):
                if i > 0 and nums[i] == nums[i-1]:
                    continue
                for sub in permute(nums[0:i:] + nums[i+1::]):
                    res.append([nums[i], *sub])
            return res
        
        return permute(nums)